src/NServiceBus.Core/Transports/Msmq/ReceiveWithTransactionScope.cs

```diff
@@ -19,45 +19,118 @@ public ReceiveWithTransactionScope(TransactionOptions transactionOptions)
-                    Logger.Error(error, ex);
+                    message = inputQueue.Receive(TimeSpan.FromMilliseconds(10), MessageQueueTransactionType.Automatic);
 
-                    errorQueue.Send(message, MessageQueueTransactionType.Automatic);
+                    if (PerformRecoverabilityAction(message))
+                    {
+                        scope.Complete();
+                        return;
+                    }
 
-                    scope.Complete();
-                    return;
-                }
+                    try
+                    {
+                        headers = MsmqUtilities.ExtractHeaders(message);
+                    }
+                    catch (Exception ex)
+                    {
+                        var error = $"Message '{message.Id}' is corrupt and will be moved to '{errorQueue.QueueName}'";
+                        Logger.Error(error, ex);
 
-                using (var bodyStream = message.BodyStream)
-                {
-                    var ambientTransaction = new TransportTransaction();
-                    ambientTransaction.Set(Transaction.Current);
-                    var pushContext = new PushContext(message.Id, headers, bodyStream, ambientTransaction, cancellationTokenSource, new ContextBag());
+                        errorQueue.Send(message, MessageQueueTransactionType.Automatic);
 
-                    await onMessage(pushContext).ConfigureAwait(false);
-                }
+                        scope.Complete();
+                        return;
+                    }
+
+                    using (var bodyStream = message.BodyStream)
+                    {
+                        var ambientTransaction = new TransportTransaction();
+                        ambientTransaction.Set(Transaction.Current);
+                        var pushContext = new PushContext(message.Id, headers, bodyStream, ambientTransaction, cancellationTokenSource, new ContextBag());
 
-                if (!cancellationTokenSource.Token.IsCancellationRequested)
+                        await onMessage(pushContext).ConfigureAwait(false);
+                    }
+
+                    if (!cancellationTokenSource.Token.IsCancellationRequested)
+                    {
+                        scope.Complete();
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                if (message == null)
                 {
-                    scope.Complete();
+                    throw;
                 }
+                var numberOfRetries = GetNumberOfRetries(message.Id);
+
+                var context = new ErrorContext(ex,
+                    message.Id,
+                    headers ?? new Dictionary<string, string>(),
+                    numberOfRetries
+                    );
+
+                var recoveryAction = await onError(context).ConfigureAwait(false);
+
+                AddToLRU(message.Id, recoveryAction);
```

Comment:
Transports like Rabbit, ASQ and ASB could just perform the action right here since they don't need to rollback (same goes for MSMQ/SQL in no tx mode)